sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -401,42 +401,51 @@ object GpuOverrides {
     !regexList.exists(pattern => s.contains(pattern))
   }
 
-  private def canonicalizeToCpuForSortOrder(exp: Expression): Expression = exp match {
-    case g: GpuLiteral => Literal(g.value, g.dataType).canonicalized
-    case g: GpuKnownFloatingPointNormalized =>
-      KnownFloatingPointNormalized(canonicalizeToCpuForSortOrder(g.child)).canonicalized
-    case g: GpuNormalizeNaNAndZero =>
-      NormalizeNaNAndZero(canonicalizeToCpuForSortOrder(g.child)).canonicalized
-    case g: GpuAlias =>
-      ShimLoader.getSparkShims.alias(canonicalizeToCpuForSortOrder(g.child), g.name)(
-        g.exprId,
-        g.qualifier,
-        g.explicitMetadata)
-          .canonicalized
-    case g: GpuSubstring =>
-      Substring(
-        canonicalizeToCpuForSortOrder(g.str),
-        canonicalizeToCpuForSortOrder(g.pos),
-        canonicalizeToCpuForSortOrder(g.len))
-          .canonicalized
-    case o: GpuExpression =>
-      throw new IllegalStateException(s"${o.getClass} is not expected to be a part of a SortOrder")
-    case other => other.canonicalized
+  private def convertExprToGpuIfPossible(expr: Expression, conf: RapidsConf): Option[Expression] = {
+    val wrapped = wrapExpr(expr, conf, None)
+    wrapped.tagForGpu()
+    if (wrapped.canExprTreeBeReplaced) {
+      Some(wrapped.convertToGpu())
+    } else {
+      None
+    }
   }
 
-  private def gpuOrderingSemanticEquals(found: Expression, required: Expression): Boolean = {
-    found.deterministic && required.deterministic &&
-        canonicalizeToCpuForSortOrder(found) == canonicalizeToCpuForSortOrder(required)
+  private def gpuOrderingSemanticEquals(
+      found: Expression,
+      required: Expression,
+      conf: RapidsConf): Boolean = {
+    if (found.deterministic && required.deterministic) {
+      val foundOnGpu = found.find(_.isInstanceOf[GpuExpression]).isDefined
+      val requiredOnGpu = required.find(_.isInstanceOf[GpuExpression]).isDefined
+      if (foundOnGpu == requiredOnGpu) {
+        found.canonicalized == required.canonicalized
+      } else if (foundOnGpu) {
```

Comment:
I think this would be simpler if we had the conversion do a best-effort conversion rather than an optional result.  That avoids the three cases that need to be handled, e.g.:
```scala
  if (found.deterministic && required.deterministic) {
    convertToGpuIfPossible(found, conf).canonicalized == convertToGpuIfPossible(required, conf).canonicalized
  }
```

and `convertToGpuIfPossible` returns an `Expression` that may or may not have been converted to a `GpuExpression`.

Reply:
I am happy to make the change.  The reason I did it this way was to avoid converting things to the GPU if they were not needed.  It might be premature optimization so if you want me to change it I will.

Reply:
I was thinking it could still avoid it using the same `find` approach being done, just moving that into `convertToGpuIfPossible`, e.g.:
```scala
def convertToGpuIfPossible(expr: Expression): Expression = {
  if (expr.find(_.isInstanceOf[GpuExpression]).isDefined) {
    // already been converted
    expr
  } else {
    val wrapped = wrapExpr(expr, conf, None)
    wrapped.tagForGpu()
    if (wrapped.canExprTreeBeReplaced) {
      wrapped.convertToGpu()
    } else {
      expr
    }
  }
}
```

Reply:
The one case that misses is if both are on the CPU already.  But it is probably not a big deal, so I'll make the change.